core/src/test/java/org/apache/iceberg/rest/TestHTTPClient.java

```diff
@@ -121,6 +128,103 @@ public void testHeadFailure() throws JsonProcessingException {
     testHttpMethodOnFailure(HttpMethod.HEAD);
   }
 
+  @Test
+  public void testProxyServer() throws IOException {
+    int proxyPort = 1070;
+    String proxyHostName = "localhost";
+    try (ClientAndServer proxyServer = startClientAndServer(proxyPort);
+        RESTClient clientWithProxy =
+            HTTPClient.builder(ImmutableMap.of())
+                .uri(URI)
+                .withProxy(proxyHostName, proxyPort)
+                .build()) {
+      String path = "v1/config";
+
+      HttpRequest mockRequest =
+          request("/" + path).withMethod(HttpMethod.HEAD.name().toUpperCase(Locale.ROOT));
+
+      HttpResponse mockResponse = response().withStatusCode(200);
+
+      mockServer.when(mockRequest).respond(mockResponse);
+      proxyServer.when(mockRequest).respond(mockResponse);
+
+      clientWithProxy.head(path, ImmutableMap.of(), (onError) -> {});
+      proxyServer.verify(mockRequest, VerificationTimes.exactly(1));
+    }
+  }
+
+  @Test
+  public void testProxyCredentialProviderWithoutProxyServerFailsBuild() throws IOException {
+    int proxyPort = 1070;
+    String proxyHostName = "localhost";
+    String authorizedUsername = "test-username";
+    String authorizedPassword = "test-password";
+    HttpHost proxy = new HttpHost(proxyHostName, proxyPort);
+    BasicCredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+    credentialsProvider.setCredentials(
+        new AuthScope(proxy),
+        new UsernamePasswordCredentials(authorizedUsername, authorizedPassword.toCharArray()));
+    Assertions.assertThatThrownBy(
+            () -> {
+              HTTPClient.builder(ImmutableMap.of())
+                  .uri(URI)
+                  .withProxyCredentialsProvider(credentialsProvider)
```

Comment:
there should also be a separate test method where the proxy hostname is set to null